🚀 3단계 - 경로 조회 기능

ktae23 opened the pull request:
알고리즘 라이브러리를 사용해서 구현해보기는 처음이라 재밌는 시간이었습니다.
이번에는 인수조건 먼저 작성하고 인사이드 아웃으로 구현해보려 했습니다.
회사에서는 도메인 테스트가 거의 없고 대부분 서비스를 모킹해서 사용하는데 도메인에서 실 객체를 활용한 테스트를 하도록 위임하고 나니 모킹이 더 번거롭다는 느낌이 드네요.
저희 서비스 코드에는 도메인보다 서비스 레이어에 불필요한 책임이 부여되지 않았나 돌아보게 됩니다.

ktae23 committed:
test : write path acceptance test case

ktae23 committed:
test : write path finder test case

ktae23 committed:
feat : develop path finder

ktae23 committed:
feat : develop path finder service

ktae23 committed:
feat : develop path finder controller

ktae23 committed:
feat : fix can add condition

testrace commented on src/main/java/nextstep/subway/domain/PathFinder.java:
```diff
@@ -0,0 +1,62 @@
+                .map(Line::getSections)
+                .flatMap(Collection::stream)
+                .forEach(section -> graph.setEdgeWeight(graph.addEdge(section.getUpStation(), section.getDownStation()), section.getDistance()));
+    }
```
repository를 활용해 모든 노선과 역을 경로 조회 대상에 포함시켜주셨네요!
PathFinder는 경로를 찾는 것에 더 집중하기 위해 경로 조회에 대상이 되는 구간 목록을 인자로 받으면 어떨까요?
경로 조회를 위한 vertex, edge, weight는 Section에 모두 포함이 되어 있으니 `List<Section>` 을 인자로 받을 수 있을 것 같아요

testrace commented on src/main/java/nextstep/subway/domain/vo/Sections.java:
```diff
@@ -23,6 +23,10 @@ public class Sections {
     public Sections() {
     }
 
+    public List<Section> getValue() {
+        return sections;
+    }
```
방어적 복사를 위해 불변 컬렉션으로 반환해도 좋을 것 같아요 😃 
```suggestion
    public List<Section> getValue() {
        return Collections.unmodifiableList(sections);
    }
```

testrace commented on src/test/java/nextstep/subway/acceptance/PathAcceptanceTest.java:
```diff
@@ -0,0 +1,109 @@
+
+    }
+
+    @Nested
+    class Fail {
```
예외 테스트 👍

testrace commented on src/test/java/nextstep/subway/unit/PathFindServiceTest.java:
```diff
@@ -0,0 +1,193 @@
+    private StationRepository stationRepository;
+
+    @Mock
+    private PathFinder pathFinder;
```
PathFinder 내부에서 Repository 에 의존하지 않는다면  mocking을 제거할 수 있을 것 같아요 😃

testrace reviewed (changes requested):
안녕하세요 경태님 😃 
커밋 로그로 흐름을 파악할 수 있어서 좋았습니다 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다시 리뷰 요청 부탁드려요!

ktae23 committed:
refactor : apply code review

1. remove bean injection at PathFinder and replace to List of Section
2. Line return un modifiable list instead origin sections value
3. modify relation entity loading strategy lazy to eager because it has problem use hibernate proxy

ktae23 commented:
PathFinder에 빈 주입 없이 처리가 가능하니까 훨씬 수월해지네요.
객체를 만들기보단 서비스를 만들고 빈으로 등록해서 처리하는게 너무 손에 익었나봐요.
제안 주신 리뷰들 모두 제가 생각 치 못했던 부분이라 다시 한번 생각을 깰 수 있어서 좋았습니다.
감사합니다!

testrace commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -34,7 +35,7 @@ private Line(String name, String color, Station upStation, Station downStation,
 
     public List<Section> getSections() {
-        return sections.getValue();
+        return Collections.unmodifiableList(sections.getValue());
```
피드백 반영 👍

testrace commented on src/main/java/nextstep/subway/domain/PathFinder.java:
```diff
@@ -50,13 +43,17 @@ private Optional<GraphPath<Station, DefaultWeightedEdge>> initializePath(Station
 
     private void addStationVertexes(WeightedMultigraph<Station, DefaultWeightedEdge> graph) {
-        stationRepository.findAll().forEach(graph::addVertex);
+        sections.stream()
+                .map(Section::getUpStation)
+                .findFirst()
+                .map(graph::addVertex);
+        sections.stream()
+                .map(Section::getDownStation)
+                .forEach(graph::addVertex);
```
stream api 의 flatMap 메서드를 활용해 보셔도 좋을 것 같아요 😃

ktae23 replied:
flatMap은 그저 결과의 형태에 따른 처리를 위해서만 사용해봤었는데요.
이렇게 먼저 flatMap을 사용해야지! 하고 코드를 바라보고 처리해보니 정말 좋네요..!
오랜만에 코드 보고 기분이 좋았습니다. ㅋㅋㅋ

감사합니다!

```java
        sections.stream()
                .map(section -> Set.of(section.getUpStation(), section.getDownStation()))
                .flatMap(Collection::stream)
                .forEach(graph::addVertex);

testrace commented on src/main/java/nextstep/subway/service/PathFindService.java:
```diff
@@ -3,22 +3,29 @@
+import java.util.List;
 import java.util.stream.Collectors;
 
+@Transactional(readOnly = true)
```
Transactional 👍

testrace commented on src/test/java/nextstep/subway/unit/PathFinderTest.java:
```diff
@@ -34,7 +38,6 @@ public class PathFinderTest {
     @Autowired
     private LineRepository lineRepository;
```
PathFinder 도메인 객체가 repository를 더 이상 활용하지 않게 되었으니 테스트에서도 제거하면 어떨까요?
엔티티의 id를 생성하기 위해 repository를 활용하기보다 대신 생성자를 추가하거나 스프링에서 제공하는 ReflectionTestUtils 를 활용해 보셔도 좋을 것 같아요

testrace reviewed (changes requested):
안녕하세요 경태님 😃 
피드백 반영 잘해주셨네요 👍 
조금만 더 다듬고 마무리하면 좋을 것 같아 코멘트 남겼습니다
확인해 주시고 다시 리뷰요청 부탁드려요

ktae23 committed:
refactor : apply code review

testrace commented on src/main/java/nextstep/subway/domain/PathFinder.java:
```diff
@@ -44,13 +42,9 @@ private Optional<GraphPath<Station, DefaultWeightedEdge>> initializePath(Station
-        sections.stream()
-                .map(Section::getDownStation)
+                .map(section -> Set.of(section.getUpStation(), section.getDownStation()))
+                .flatMap(Collection::stream)
```
피드백 반영 👍

testrace commented on src/test/java/nextstep/subway/unit/PathFinderTest.java:
```diff
@@ -6,20 +6,19 @@
 import static org.assertj.core.api.BDDAssertions.thenCode;
+import static org.mockito.Mockito.spy;
 
 @ClassicUnitTest
```
SpringBootTest를 포함한 애너테이션은 제거해도 괜찮을 것 같아요 😃 
```suggestion

```

testrace reviewed (approved):
안녕하세요 경태님 😃 
피드백 반영 잘해주셨네요 👍 
2주차 미션은 여기서 마무리하겠습니다
남은 미션도 화이팅입니다 🔥
